### ISSUE TITLE
What is the advantage of using `gluonnlp.utils.parallel`?

### ISSUE THREAD

**Issue by zeeshansayyed** [dmlc/gluon-nlp#999]
**Question about the library**

The machine translation example with [transformers](https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/train_transformer.py) uses [parallel](https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/utils/parallel.py). How is this different from the [automatic parallelism ](http://d2l.ai/chapter_computational-performance/auto-parallelism.html) which is built into MxNet? When should one use it?

Thanks

**Comment by eric-haibin-lin** [comment]
When you got a really fast GPU, a single python thread could be relatively slow to push computation to all GPUs when the batch size is small (GPU computation time is short). The parallel API is an attempt to improve that. If training time is not your bottleneck, you don't need to use it

**Comment by zeeshansayyed** [comment]
Thanks. This makes sense. I guess the best way to know whether you would need it is to actually implement both and see whether you are gaining any performance. Other than that, are there any rules of thumbs on GPU type and batch size which can help me decide this without actually implementing both? I am using V100s available on EC2.

**Comment by eric-haibin-lin** [comment]
You're right. We might work on a simpler API that don't require users to inherit from the interface so that users only need to change 1 or 2 lines of code. But not yet started. 
What is the network architecture and model parameter size? Is your model a hybrid block?

**Comment by zeeshansayyed** [comment]
My network is on the bigger side. I have BERT as the encoder (which is hybrid), but the overall network is not hybrid.

**Comment by feevos** [comment]
Hi @eric-haibin-lin  is it possible to use the parallel module (modified, including only forward) to do a parallel forward pass? What I need to do is something like (demo pseudo-code):

```Python
class SomeNet(HybridBlock):
    def __input__(self, SomeArguments):
         super().__init__(self)

         net = gluon.HybridSequential()
         for _ in range(5):
             net.add(gluon.nn.Conv2D(channels=32,kernel_size=3,padding=1))
        self.convs = net
    def forward(self, *inputs):
        # Here inputs is a list of mx.np.array() of different spatial size 
        outs = []
        for input, conv in zip(inputs,self.convs):
            outs = outs + [conv(input)]
        return outs 

```

I just found your answer and I will test it, but I would appreciate your expert advice. 

Regards,
Foivos

**Comment by szha** [comment]
@feevos you can adapt the parallel function for the purpose of pushing work to GPU more efficiently. The main purpose of that interface is to push work with multiple threads using multithreaded queues. See https://github.com/dmlc/gluon-nlp/blob/v0.x/src/gluonnlp/utils/parallel.py#L125-L139. For forward-only, you just need to implement only the forward part, in which case autograd.record isn't needed.

**Comment by feevos** [comment]
Thank you very much @szha !!!
